apps/prairielearn/src/lib/course-instance-access.ts

```diff
@@ -0,0 +1,112 @@
+import { Temporal } from '@js-temporal/polyfill';
+
+import { type AccessRuleJson, type PublishingJson } from '../schemas/infoCourseInstance.js';
+
+import { type CourseInstanceAccessRule } from './db-types.js';
+
+export interface PublishingConfigurationMigrationResult {
+  success: true;
+  publishingConfiguration: PublishingJson;
+}
+
+export interface PublishingConfigurationMigrationError {
+  success: false;
+  error: string;
+}
+
+export type PublishingConfigurationMigrationResponse =
+  | PublishingConfigurationMigrationResult
+  | PublishingConfigurationMigrationError;
+
+const toIsoString = (date: Date, timezone: string) => {
+  return Temporal.Instant.fromEpochMilliseconds(date.getTime())
+    .toZonedDateTimeISO(timezone)
+    .toPlainDateTime()
+    .toString();
+};
+
+/**
+ * Converts a database CourseInstanceAccessRule to the AccessRuleJson format.
+ */
+export function convertAccessRuleToJson(
+  accessRule: CourseInstanceAccessRule,
+  courseInstanceTimezone: string,
+): AccessRuleJson {
+  const json: AccessRuleJson = {};
+
+  if (accessRule.json_comment) {
+    json.comment = accessRule.json_comment;
+  }
+
+  if (accessRule.uids && accessRule.uids.length > 0) {
+    json.uids = accessRule.uids;
+  }
+
+  if (accessRule.start_date) {
+    json.startDate = toIsoString(accessRule.start_date, courseInstanceTimezone);
+  }
+
+  if (accessRule.end_date) {
+    json.endDate = toIsoString(accessRule.end_date, courseInstanceTimezone);
+  }
+
+  if (accessRule.institution) {
+    json.institution = accessRule.institution;
+  }
+
+  return json;
+}
+
+/**
+ * Attempts to migrate legacy access rules (in AccessRuleJson format) to the new publishing configuration format.
+ * Uses the earliest start date and latest end date to determine the publishing configuration.
+ */
+export function migrateAccessRuleJsonToPublishingConfiguration(
+  originalAccessRules: AccessRuleJson[],
+): PublishingConfigurationMigrationResponse {
```

Comment:
Legacy access rules support a) lists of UIDs and b) specific institutions. This will silently drop them. Did you consider handling those cases?

Reply:
Discussed with @reteps: we want to add warnings for things that we can't return. He's going to remove this function from this branch and implement it + the changes in the UI branch.